graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,156 @@
 
-    if is_uuid4_type(schema_info, pagination_node.name.value, pagination_field):
-        pass
-    elif is_int_field_type(schema_info, pagination_node.name.value, pagination_field):
-        quantiles = schema_info.pagination_keys.statistics.get_field_quantiles(
-            pagination_node, pagination_field
+    if is_uuid4_type(schema_info, pagination_node, pagination_field):
+        return (
+            PaginationPlan(
+                (VertexPartition((pagination_node,), pagination_field, number_of_pages),)
+            ),
+            [],
         )
-        # We make sure there's more than enough quantiles because we don't interpolate.
-        if quantiles is None or len(quantiles) < 10 * number_of_pages:
-            return None
-    else:
-        return None
+    elif field_supports_range_reasoning(schema_info, pagination_node, pagination_field):
+        quantiles = schema_info.statistics.get_field_quantiles(pagination_node, pagination_field)
+
+        # If the ratio len(quantiles) / number_of_pages is N, then the largest page is
+        # expected to be (N + 1) / N times larger than the smallest page. So if N is 5,
+        # (N + 1) / N = 1.2, which is not too bad.
+        ideal_quantile_resolution = 5 * number_of_pages + 1
 
-    return PaginationPlan([VertexPartition((pagination_node.name.value,), number_of_pages)])
+        if quantiles is None:
+            return (
+                PaginationPlan(tuple()),
+                [
+                    NotEnoughQuantiles(
+                        pagination_node, pagination_field, 0, ideal_quantile_resolution
+                    )
+                ],
+            )
+        elif len(quantiles) - 1 < number_of_pages:
+            return (
+                PaginationPlan(
+                    (VertexPartition((pagination_node,), pagination_field, len(quantiles) - 1),)
+                ),
+                [
+                    NotEnoughQuantiles(
+                        pagination_node, pagination_field, len(quantiles), ideal_quantile_resolution
+                    )
+                ],
+            )
+        elif len(quantiles) < ideal_quantile_resolution:
+            return (
+                PaginationPlan(
+                    (VertexPartition((pagination_node,), pagination_field, number_of_pages),)
+                ),
+                [
+                    NotEnoughQuantiles(
+                        pagination_node, pagination_field, len(quantiles), ideal_quantile_resolution
+                    )
+                ],
+            )
+        else:
+            return (
+                PaginationPlan(
+                    (VertexPartition((pagination_node,), pagination_field, number_of_pages),)
+                ),
+                [],
+            )
+    else:
+        type_name = (
+            schema_info.schema.get_type(pagination_node.name.value)
+            .fields[pagination_field]
+            .type.name
+        )
+        raise AssertionError(
```

Comment:
I'm not sure why this is an error rather than returning a no-pagination plan accompanied by an appropriate warning (an advisory, with my suggested nomenclature).

Reply:
The schema info was not initialized properly, if we designated a pagination field with unsupported type. This is on us.

Reply:
Can we make the error message say that the schema info is invalid and give some advice on how to correct it then? I would not have figured this out on the basis of the error message.